### PR TITLE
Implement wildcard analyzer

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

Using ngram to approximate wildcard queries

For 11 * 1e6 strings, with average length 800 and pattern like `"%fjdkajfkajdgagh%"`:
|cold | hot | memory | analyzer|
|----| ----| ---|---|
| \>= 200s | 35s | 8.3Gb | identity |
| 7ms | 6ms | 20.9Gb | wildcard 3 (position) |
| 20ms | < 1ms | 21.3Gb | wildcard 4 (position) |
| 25ms | < 1ms | 12.9Gb | wildcard 5 (without position) |

csd -- 8.5 Gb, can be compressed...

Can be done in the future:
1. Avoiding additional compute for positions
2. Implement another analyzer without column.
3. Allowing reads from analyzer column
4. Optimize "in" for such field (disjunction of ngram_matches)

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: tests/js/client/aql/aql-arangosearch-wildcard-analyzer.js [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      let res = query("abcdef", false, !hasPos);
+      assertEqual(res, ["abcdef"]);
+    },
+
+    testAnyShort: function() {
+      let res = query("_", false, true);
+      assertEqual(res, ["a", "c", "e", "f"]);
+    },
+
+    testAnyMid: function() {
```
Comment (Dronplane) [inline tests/js/client/aql/aql-arangosearch-wildcard-analyzer.js:114]: We also need "_xxxx" pattern and "xxx%xxx"

File: arangod/IResearch/Wildcard/Options.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    token = best;
+  } else {
+    hasPos = analyzer.features().hasFeatures(irs::IndexFeatures::POS);
+  }
+  if (needsMatcher || !hasPos) {
+    matcher = ctx->buildLikeMatcher(pattern, true);
+  }
+}
+
+}  // namespace arangodb::iresearch::wildcard
```
Comment (Dronplane) [inline arangod/IResearch/Wildcard/Options.cpp:101]: CR

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 3]
```diff
@@ -3113,9 +3118,14 @@ Result Features::validate() const {
             "specified"};
   }
 
-  constexpr irs::IndexFeatures kSupportedFeatures = irs::IndexFeatures::OFFS |
-                                                    irs::IndexFeatures::POS |
-                                                    irs::IndexFeatures::FREQ;
+  // wildcard shouldn't have offs in the index
+  // TODO(MBkkt) geo analyzers shouldn't have freq, pos, offs in the index
+  bool isWildcardAnalyzer = type == wildcard::Analyzer::type_name();
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp:3123]: I think `==` is more clear

File: arangod/IResearch/Wildcard/Analyzer.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                               irs::slice_to_string(input), kParseError));
+    return false;
+  }
+  return true;
+}
+
+static constexpr std::string_view kFill = "00000\xFF";
+
+constexpr std::string_view fill(size_t len) noexcept {
+  return {kFill.data() + kFill.size() - len, len};
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/IResearch/Wildcard/Analyzer.cpp:102]: It's kind of compile time, so no

File: arangod/IResearch/Wildcard/Analyzer.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+bool Analyzer::reset(std::string_view data) {
+  if (!_analyzer->reset(data)) {
+    return false;
+  }
+  _terms.clear();
+  while (_analyzer->next()) {
+    auto const size = _term->value.size();
+    if (size > std::numeric_limits<int32_t>::max()) {
+      // TODO(MBkkt) icu doesn't support more
+      LOG_TOPIC("14345", ERR, TOPIC)
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/IResearch/Wildcard/Analyzer.cpp]: Done

File: arangod/IResearch/Wildcard/Analyzer.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+ private:
+  irs::byte_type* begin() noexcept {
+    return reinterpret_cast<irs::byte_type*>(_terms.data());
+  }
+
+  analyzer::ptr _analyzer;
+  std::unique_ptr<Ngram> _ngram;
+  irs::term_attribute const* _term{};
+  irs::term_attribute* _ngramTerm{};
+  irs::bytes_view _lastNgram;
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/IResearch/Wildcard/Analyzer.h]: Done

File: arangod/IResearch/Wildcard/Analyzer.cpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+                                velocypack::Slice slice) {
+  auto& impl = basics::downCast<Analyzer>(*ctx);
+  return irs::ViewCast<irs::byte_type>(std::string_view{impl._terms});
+}
+
+bool Analyzer::next() {
+  if (IRS_LIKELY(_ngram->next())) {
+    return true;
+  }
+  if (auto size = _ngramTerm->value.size(); size > 1) {
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/IResearch/Wildcard/Analyzer.cpp:192]: I don't understand

[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/IResearch/Wildcard/Analyzer.cpp:192]: Suggest code or I will ignore it

File: arangod/IResearch/Wildcard/Options.cpp (hunk 2 above)
Comment (MBkkt) [reply arangod/IResearch/Wildcard/Options.cpp]: Done

File: arangod/IResearch/Wildcard/Analyzer.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dronplane) [reply arangod/IResearch/Wildcard/Analyzer.cpp:192]: Something like that (not exact suggestion but more a prototype):
`
auto* offset = irs::get<irs::offset>(_ngram);
uint32_t lastStart{0};
bool next() {
  if (IRS_LIKELY(_ngram->next())) {
    lastStart = offset.start;
    return true;
  }
  if (lastStart != data.size()) {
    auto* begin = data.data()+ lastStart;
    auto* end = data.data() + data.size();
    begin = nextUTF8(begin, end);
    lastStart = begin - data.data();
    return true;
  }
 ... 
   if (!_ngram->next()) {
    _ngramTerm->value = term;
  }
  lastStart = offset.start;
  return true;
}

`

[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/IResearch/Wildcard/Analyzer.cpp:192]: I think it's more complicated

**Review by Dronplane — APPROVED** [review]
LGTM

**Pull request merged**